net.bobbo.player-controller/player_scripts/crouch.gd

```diff
@@ -14,30 +14,60 @@ var crouch_current: float = 0
 var crouch_target: float = 0
 const crouch_action: String = "player_crouch"
 
-## default_player_height is needed for proper uncrouching height calculations 
+## _default_player_height is needed for proper uncrouching height calculations 
 ## if the maximum player height were to change during gameplay at any point this 
 ## value would need to somehow be updated to account for that default is 2 
 ## because the default height in player_controller is 2
 var _default_player_height: float = 2
-var _should_be_crouched: bool = false
+
+var _ray_cast: RayCast3D = null
 
 const _standing_height_multiplier: float = 1
 const _crouching_height_multiplier: float = 0.5
 
 #
 #	Functions
-#
+# 
 
 func player_ready():
 	_default_player_height = player.height
+	
+	_ray_cast = RayCast3D.new()
+	_ray_cast.add_exception(player)
+	
+	# I split this up into a variable to keep it behind the second line
+	var inital_target_position = Vector3(0, _default_player_height - player.position.y, 0)
+	_ray_cast.target_position = player.up_direction * inital_target_position
+	add_child(_ray_cast)
+	
 	self.assert_input_action(crouch_action)
 
 func player_physics_process(delta: float):
-	if Input.is_action_pressed(crouch_action) or _should_force_crouch():
+	# Calculate origin and target position of raycast
+	_ray_cast.position = player.position
```

Comment:
I believe this will always make the ceiling cast happen a frame behind because we manually have to update it's position every tick...